packages/foundry/contracts/hooks/SwapDiscountHook/SwapDiscountHook.sol

```diff
@@ -0,0 +1,119 @@
+    IDiscountCampaignFactory public discountCampaignFactory;
+    address private immutable allowedFactoryAddress;
+    address private immutable trustedRouterAddress;
+    uint256 private _shareTokenId = 1;
+
+    constructor(
+        IVault vaultInstance,
+        address _factoryAddress,
+        address _routerAddress,
+        address _campaignFactory,
+        string memory name,
+        string memory symbol
+    ) VaultGuard(vaultInstance) ERC721(name, symbol) Ownable(msg.sender) {
+        allowedFactoryAddress = _factoryAddress;
+        trustedRouterAddress = _routerAddress;
+        discountCampaignFactory = IDiscountCampaignFactory(_campaignFactory);
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(
+        address factory,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public view override returns (bool) {
+        return factory == allowedFactoryAddress && IBasePoolFactory(factory).isPoolFromFactory(pool);
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory hookFlags) {
+        hookFlags.shouldCallAfterSwap = true;
+        return hookFlags;
+    }
+
+    /// @inheritdoc IHooks
+    function onAfterSwap(
+        AfterSwapParams calldata params
+    ) public override onlyVault returns (bool success, uint256 discountedAmount) {
+        (address campaignAddress, ) = discountCampaignFactory.discountCampaigns(params.pool);
+        if (campaignAddress != address(0)) {
+            IDiscountCampaign campaign = IDiscountCampaign(campaignAddress);
+            (bytes32 campaignID, , , , , address rewardToken, address poolAddress, ) = campaign.campaignDetails();
+            if (
+                params.kind == SwapKind.EXACT_IN &&
+                address(params.tokenOut) == rewardToken &&
+                poolAddress == params.pool
+            ) {
+                mint(params, campaign, campaignID);
+            }
+        }
+        return (true, params.amountCalculatedRaw);
+    }
+
+    /**
+     * @notice Apply discount and mint a new token for the user.
+     * @dev This function mints a token for the user after applying the discount and storing relevant swap data.
+     *      The function uses memory for `UserSwapData` to minimize storage operations until the final write to storage.
+     * @param params The parameters of the swap after it has been executed.
+     *        - params.router: The address of the router handling the swap.
+     *        - params.amountCalculatedRaw: The amount swapped by the user.
+     *        - params.pool: The liquidity pool involved in the swap.
+     */
+    function mint(
+        AfterSwapParams calldata params,
+        IDiscountCampaign _campaign,
+        bytes32 _campaignID
+    ) internal nonReentrant {
+        uint256 newTokenId = _shareTokenId++;
+        address user = IRouterCommon(params.router).getSender();
```

Comment:
@EndymionJkb I've added the support for the router check but how can I add the support for batch routers?

Reply:
You could just pass in something like `trustedSimpleRouter` and `trustedBatchRouter`, since there are only two, and check that the router address is one or the other.

Or you could be more general, and use our EnumerableSet library to store a set of them, and have permissioned functions to add/remove from the set (e.g., in case we deprecate and replace a router, which is possible - otherwise, you'd have to redeploy everything if the router changed).